libcoap/src/types.rs

```diff
@@ -694,13 +694,28 @@ impl CoapUri {
     pub unsafe fn from_raw_uri(raw_uri: *const coap_uri_t, is_proxy: bool) -> CoapUri {
         // Loosely based on coap_clone_uri.
         assert!(!raw_uri.is_null());
+        let host_slice = if (*raw_uri).host.s.is_null() {
+            &[]
+        } else {
+            std::slice::from_raw_parts((*raw_uri).host.s, (*raw_uri).host.length)
+        };
+        let path_slice = if (*raw_uri).path.s.is_null() {
+            &[]
+        } else {
+            std::slice::from_raw_parts((*raw_uri).path.s, (*raw_uri).path.length)
+        };
+        let query_slice = if (*raw_uri).query.s.is_null() {
+            &[]
+        } else {
+            std::slice::from_raw_parts((*raw_uri).query.s, (*raw_uri).query.length)
+        };
```

Comment:
Alternative:
```suggestion
        let host_slice = (*raw_uri).host.s.is_null().then_some(&[]).unwrap_or_else(|| std::slice::from_raw_parts((*raw_uri).host.s, (*raw_uri).host.length));
        let path_slice = (*raw_uri).path.s.is_null().then_some(&[]).unwrap_or_else(|| std::slice::from_raw_parts((*raw_uri).path.s, (*raw_uri).path.length));
        let query_slice = (*raw_uri).query.s.is_null().then_some(&[]).unwrap_or_else(|| std::slice::from_raw_parts((*raw_uri).query.s, (*raw_uri).query.length));
```

But I'm not sure if this is necessarily better than the way it is right now.